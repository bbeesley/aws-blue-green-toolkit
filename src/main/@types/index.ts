export * from './aurora-tools.js';
export * from './cloudwatch-tools.js';
export * from './common.js';
export * from './dynamo-tools.js';
export * from './kinesis-tools.js';
export * from './lambda-tools.js';
export * from './sns-tools.js';
export * from './sqs-tools.js';
export * from './ecs-tools.js';
