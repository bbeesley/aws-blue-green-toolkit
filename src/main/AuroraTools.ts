import { RDS, ApplicationAutoScaling } from 'aws-sdk';
import { SNSEventRecord } from 'aws-lambda';
import { AwsConfig } from './common-interfaces';
import { DatabaseReference, ClusterState } from './constants';

/**
 * Configuration options for the Aurora toolkit
 * @interface AuroraConfig
 * @extends {AwsConfig}
 */
interface AuroraConfig extends AwsConfig {
  clusterNameA: string;
  clusterNameB: string;
  minimumClusterSize: number;
  skipDeleteSnapshots: boolean;
}

/**
 * Toolkit for Aurora operations
 * @export
 * @class AuroraTools
 */
export class AuroraTools {
  config: AuroraConfig;
  rds: RDS;
  aas: ApplicationAutoScaling;

  /**
   *Creates an instance of AuroraTools.
   * @param {AuroraConfig} config - Configuration options for the Aurora toolkit
   * @memberof AuroraTools
   */
  public constructor(config: AuroraConfig) {
    this.config = config;
    this.rds = new RDS({ region: this.config.awsRegion });
    this.aas = new ApplicationAutoScaling({ region: this.config.awsRegion });
  }

  /**
   * Returns the cluster name associated with a database reference
   * @protected
   * @param {DatabaseReference} ref - Reference to a db cluster
   * @returns {string} - The name of the db cluster
   * @memberof AuroraTools
   */
  protected getClusterName(ref: DatabaseReference): string {
    return ref === DatabaseReference.a
      ? this.config.clusterNameA
      : this.config.clusterNameB;
  }

  /**
   * Returns the cluster name of a cluster's partner cluster
   * @protected
   * @param {DatabaseReference} ref - Reference to a db cluster
   * @returns {string} - The name of the partner cluster
   * @memberof AuroraTools
   */
  protected getClusterPartnerName(ref: DatabaseReference): string {
    return ref === DatabaseReference.b
      ? this.config.clusterNameA
      : this.config.clusterNameB;
  }

  /**
   * Returns a partner reference, given a reference
   * @protected
   * @param {DatabaseReference} ref - Reference to a db cluster
   * @returns {DatabaseReference} - The reference to the partner
   * @memberof AuroraTools
   */
  protected getClusterPartnerRef(ref: DatabaseReference): DatabaseReference {
    return ref === DatabaseReference.a
      ? DatabaseReference.b
      : DatabaseReference.a;
  }

  /**
   * Gets the current state of one of the Aurora clusters
   * @param {DatabaseReference} reference - Reference to a db cluster
   * @returns {Promise<ClusterState>}
   * @memberof AuroraTools
   */
  public async getClusterState(
    reference: DatabaseReference
  ): Promise<ClusterState> {
    const DBClusterIdentifier = this.getClusterName(reference);
    const { DBClusters } = await this.rds
      .describeDBClusters({ DBClusterIdentifier })
      .promise();
    if (DBClusters && DBClusters.length === 1) {
      const { Status } = DBClusters[0];
      if (Status) {
        if (Status === ClusterState.AVAILABLE) return ClusterState.AVAILABLE;
        if (Status === ClusterState.STOPPED) return ClusterState.STOPPED;
        if (Status === ClusterState.STARTING) return ClusterState.STARTING;
        if (Status === ClusterState.STOPPING) return ClusterState.STOPPING;
      }
    }
    console.error(`unable to get cluster info for ${DBClusterIdentifier}`);
    throw new Error(`unable to get cluster info for ${DBClusterIdentifier}`);
  }

  /**
   * Scale a database cluster to a given minimum capacity
   * @protected
   * @param {DatabaseReference} reference - Reference to a db cluster
   * @param {number} minCapacity - The desired minimum capacity to set
   * @returns {Promise<ApplicationAutoScaling.RegisterScalableTargetResponse>}
   * @memberof AuroraTools
   */
  protected async scale(
    reference: DatabaseReference,
    minCapacity: number
  ): Promise<ApplicationAutoScaling.RegisterScalableTargetResponse> {
    const db = this.getClusterName(reference);
    const aasResponse = await this.aas
      .describeScalableTargets({
        ServiceNamespace: 'rds',
        ResourceIds: [`cluster:${db}`],
      })
      .promise();
    if (
      aasResponse &&
      aasResponse.ScalableTargets &&
      aasResponse.ScalableTargets[0]
    ) {
      const scalingTarget = aasResponse.ScalableTargets[0];
      const { ServiceNamespace, ResourceId, ScalableDimension } = scalingTarget;
      console.log(
        `setting minimum capacity of ${ResourceId} cluster to ${minCapacity}`
      );
      return await this.aas
        .registerScalableTarget({
          ResourceId,
          ScalableDimension,
          ServiceNamespace,
          MinCapacity: minCapacity,
        })
        .promise();
    }
    throw new Error(`unable to scale out db cluster ${db}`);
  }

  /**
   * Reverts a cluster's minimum reader count to the configured minimum
   * @param {DatabaseReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async scaleIn(reference: DatabaseReference): Promise<void> {
    await this.scale(reference, this.config.minimumClusterSize);
  }

  /**
   * Scales out a cluster to match it's partner's size
   * @param {DatabaseReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async scaleOut(reference: DatabaseReference): Promise<void> {
    const desiredSize = await this.getReaderCount(
      this.getClusterPartnerRef(reference)
    );
    await this.scale(reference, desiredSize);
  }

  /**
   * Get a count of the number of active readers for a cluster
   * @param {DatabaseReference} reference - Reference to a db cluster
   * @returns {Promise<number>} - The number of active readers
   * @memberof AuroraTools
   */
  public async getReaderCount(reference: DatabaseReference): Promise<number> {
    const DBClusterIdentifier = this.getClusterName(reference);
    const clusterDescription = await this.rds
      .describeDBClusters({ DBClusterIdentifier })
      .promise();
    if (
      clusterDescription &&
      clusterDescription.DBClusters &&
      clusterDescription.DBClusters[0] &&
      clusterDescription.DBClusters[0].DBClusterMembers
    ) {
      return clusterDescription.DBClusters[0].DBClusterMembers.length - 1;
    }
    throw new Error('unable to count active db readers');
  }

  /**
   * Starts a stopped db cluster
   * @param {DatabaseReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async startDatabase(reference: DatabaseReference): Promise<void> {
    const DBClusterIdentifier = this.getClusterName(reference);
    await this.rds.startDBCluster({ DBClusterIdentifier }).promise();
  }

  /**
   * Stops a running db cluster
   * @param {DatabaseReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async stopDatabase(reference: DatabaseReference): Promise<void> {
    const DBClusterIdentifier = this.getClusterName(reference);
    await this.rds.stopDBCluster({ DBClusterIdentifier }).promise();
  }

  /**
   * Deletes a running db cluster
   * @param {DatabaseReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async deleteDatabase(reference: DatabaseReference): Promise<void> {
    const DBClusterIdentifier = this.getClusterName(reference);
    const { DBClusters } = await this.rds
      .describeDBClusters({ DBClusterIdentifier })
      .promise();
    if (DBClusters && DBClusters.length === 1) {
      const { DBClusterMembers } = DBClusters[0];
      if (DBClusterMembers && DBClusterMembers.length > 0) {
        await Promise.all(
          DBClusterMembers.map(async ({ DBInstanceIdentifier }) => {
            if (DBInstanceIdentifier) {
              await this.rds
                .deleteDBInstance({
                  DBInstanceIdentifier,
                  SkipFinalSnapshot: true,
                })
                .promise();
            }
          })
        );
        await this.rds
          .deleteDBCluster({
            DBClusterIdentifier,
            SkipFinalSnapshot: this.config.skipDeleteSnapshots,
          })
          .promise();
      } else {
        throw new Error(`cluster ${DBClusterIdentifier} has no members`);
      }
    } else {
      throw new Error(`unable to get cluster info for ${DBClusterIdentifier}`);
    }
  }

  /**
   * Parses a message from an rds event subscription, if the event was triggered by a scale out
   * operation, the tags defined in config are applied to the newly created reader.
   * @param {SNSEventRecord} record - An SNS event record of the type published by rds event streams
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async applyTags(record: SNSEventRecord): Promise<void> {
    const message = JSON.parse(record.Sns.Message);
    if (message['Event Message'] === 'DB instance created') {
      const Tags = Object.entries(this.config.tags).map(([Key, Value]) => ({
        Key,
        Value,
      }));
      Tags.push({ Key: 'Role', Value: 'Reader' });
      const ResourceName = `arn:aws:rds:${this.config.awsRegion}:${this.config.awsProfile}:db:${message['Source ID']}`;
      const details = await this.rds
        .listTagsForResource({
          ResourceName,
        })
        .promise();
      const scalingResourceTag = details.TagList
        ? details.TagList.find(
            t =>
              t.Key === 'application-autoscaling:resourceId' &&
              t.Value &&
              t.Value.includes(this.config.namespace)
          )
        : undefined;
      if (scalingResourceTag) {
        await this.rds
          .addTagsToResource({
            ResourceName,
            Tags,
          })
          .promise();
      }
    }
  }
}
