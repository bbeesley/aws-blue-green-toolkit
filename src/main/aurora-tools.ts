import {
  ApplicationAutoScalingClient,
  DescribeScalableTargetsCommand,
  RegisterScalableTargetCommand,
  type RegisterScalableTargetResponse,
} from '@aws-sdk/client-application-auto-scaling';
import {
  AddTagsToResourceCommand,
  DeleteDBClusterCommand,
  DeleteDBInstanceCommand,
  DescribeDBInstancesCommand,
  DescribeDBClustersCommand,
  ListTagsForResourceCommand,
  RDSClient,
  StartDBClusterCommand,
  StopDBClusterCommand,
  ModifyDBInstanceCommand,
  RDSServiceException,
  type Tag,
} from '@aws-sdk/client-rds';
import type { SNSEventRecord } from 'aws-lambda';
import delay from 'delay';
import type { AuroraConfig } from './@types/index.js';
import { ClusterState, StackReference } from './constants.js';

/**
 * Toolkit for Aurora operations
 *
 * @export
 * @class AuroraTools
 */
export class AuroraTools {
  rds: RDSClient;

  aas: ApplicationAutoScalingClient;

  /**
   * Creates an instance of AuroraTools.
   *
   * @param {AuroraConfig} config - Configuration options for the Aurora toolkit
   * @memberof AuroraTools
   */
  public constructor(public config: AuroraConfig) {
    this.config = config;
    this.rds = new RDSClient({ region: this.config.awsRegion });
    this.aas = new ApplicationAutoScalingClient({
      region: this.config.awsRegion,
    });
  }

  /**
   * Gets the current state of one of the Aurora clusters
   *
   * @param {StackReference} reference - Reference to a db cluster
   * @returns {Promise<ClusterState>}
   * @memberof AuroraTools
   */
  public async getClusterState(
    reference: StackReference
  ): Promise<ClusterState> {
    const DBClusterIdentifier = this.getClusterName(reference);
    const { DBClusters } = await this.rds.send(
      new DescribeDBClustersCommand({ DBClusterIdentifier })
    );
    if (DBClusters && DBClusters.length === 1) {
      const { Status } = DBClusters[0] ?? {};
      if (Status) {
        if (Status === ClusterState.AVAILABLE) return ClusterState.AVAILABLE;
        if (Status === ClusterState.STOPPED) return ClusterState.STOPPED;
        if (Status === ClusterState.STARTING) return ClusterState.STARTING;
        if (Status === ClusterState.STOPPING) return ClusterState.STOPPING;
      }
    }

    throw new Error(`unable to get cluster info for ${DBClusterIdentifier}`);
  }

  /**
   * Reverts a cluster's minimum reader count to the configured minimum
   *
   * @param {StackReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async scaleIn(reference: StackReference): Promise<void> {
    await this.scale(reference, this.config.minimumClusterSize);
  }

  /**
   * Scales out a cluster to match it's partner's size
   *
   * @param {StackReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async scaleOut(reference: StackReference): Promise<void> {
    const desiredSize = await this.getReaderCount(
      this.getClusterPartnerRef(reference)
    );
    await this.scale(reference, desiredSize);
  }

  /**
   * Get a count of the number of active readers for a cluster
   *
   * @param {StackReference} reference - Reference to a db cluster
   * @returns {Promise<number>} - The number of active readers
   * @memberof AuroraTools
   */
  public async getReaderCount(reference: StackReference): Promise<number> {
    const DBClusterIdentifier = this.getClusterName(reference);
    const clusterDescription = await this.rds.send(
      new DescribeDBClustersCommand({ DBClusterIdentifier })
    );
    if (clusterDescription?.DBClusters?.[0]?.DBClusterMembers) {
      return clusterDescription.DBClusters[0].DBClusterMembers.length - 1;
    }

    throw new Error('unable to count active db readers');
  }

  /**
   * Starts a stopped db cluster
   *
   * @param {StackReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async startDatabase(reference: StackReference): Promise<void> {
    const DBClusterIdentifier = this.getClusterName(reference);
    await this.rds.send(new StartDBClusterCommand({ DBClusterIdentifier }));
  }

  /**
   * Stops a running db cluster
   *
   * @param {StackReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async stopDatabase(reference: StackReference): Promise<void> {
    const DBClusterIdentifier = this.getClusterName(reference);
    await this.rds.send(new StopDBClusterCommand({ DBClusterIdentifier }));
  }

  /**
   * Deletes a running db cluster
   *
   * @param {StackReference} reference - Reference to a db cluster
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async deleteDatabase(reference: StackReference): Promise<void> {
    const DBClusterIdentifier = this.getClusterName(reference);
    const { DBClusters } = await this.rds.send(
      new DescribeDBClustersCommand({ DBClusterIdentifier })
    );
    if (DBClusters && DBClusters.length === 1) {
      const { DBClusterMembers } = DBClusters[0] ?? {};
      if (DBClusterMembers && DBClusterMembers.length > 0) {
        await Promise.all(
          DBClusterMembers.map(async ({ DBInstanceIdentifier }) => {
            if (DBInstanceIdentifier) {
              await this.rds.send(
                new DeleteDBInstanceCommand({
                  DBInstanceIdentifier,
                  SkipFinalSnapshot: true,
                })
              );
            }
          })
        );
        await this.rds.send(
          new DeleteDBClusterCommand({
            DBClusterIdentifier,
            SkipFinalSnapshot: this.config.skipDeleteSnapshots,
          })
        );
      } else {
        throw new Error(`cluster ${DBClusterIdentifier} has no members`);
      }
    } else {
      throw new Error(`unable to get cluster info for ${DBClusterIdentifier}`);
    }
  }

  /**
   * Parses a message from an rds event subscription, if the event was triggered by a scale out
   * operation, the tags defined in config are applied to the newly created reader.
   *
   * @param {SNSEventRecord} record - An SNS event record of the type published by rds event streams
   * @returns {Promise<void>}
   * @memberof AuroraTools
   */
  public async applyTags(record: SNSEventRecord): Promise<void> {
    const message = JSON.parse(record.Sns.Message) as Record<string, string>;
    if (
      message['Event Message'] === 'DB instance created' &&
      message['Source ID']
    ) {
      const ResourceName = `arn:aws:rds:${this.config.awsRegion}:${this.config.awsProfile}:db:${message['Source ID']}`;
      const details = await this.rds.send(
        new ListTagsForResourceCommand({
          ResourceName,
        })
      );
      const scalingResourceTag = details.TagList
        ? details.TagList.find(
            (t) =>
              t.Key === 'application-autoscaling:resourceId' &&
              t.Value &&
              (t.Value.includes(this.config.namespace) ||
                this.config.namespace === '*')
          )
        : undefined;
      if (scalingResourceTag) {
        const Tags = Object.entries(this.config.tags).map(([Key, Value]) => ({
          Key,
          Value,
        }));
        if (Tags.length === 0) {
          const clusterTags = await this.getTagsForCluster(ResourceName);
          Tags.push(...clusterTags);
        }

        Tags.push({ Key: 'Role', Value: 'Reader' });
        await this.rds.send(
          new AddTagsToResourceCommand({
            ResourceName,
            Tags,
          })
        );
      }
    }
  }

  /**
   * Parses a message from an rds event subscription, if the event was triggered by a scale out
   * operation and the new instance does not have performance insights enabled, the instance is updated
   * to enable performance insights.
   *
   * @param {SNSEventRecord} record - An SNS event record of the type published by rds event streams
   * @param {boolean} [reEnableIfDisabled=true] - Whether or not to automatically re enable insights if they are disabled
   * @param {number} [retryDelay=60e3] - Time in ms to wait before retrying
   * @param {number} [retryAttempts=5] - Number of retry attempts
   * @returns {*}  {Promise<void>}
   * @memberof AuroraTools
   */
  public async enablePerformanceInsights(
    record: SNSEventRecord,
    reEnableIfDisabled = true,
    retryDelay = 60e3,
    retryAttempts = 5
  ): Promise<void> {
    const message = JSON.parse(record.Sns.Message) as Record<string, string>;
    if (
      (message['Event Message'] === 'DB instance created' ||
        (reEnableIfDisabled &&
          message['Event Message'] ===
            'Performance Insights has been disabled')) &&
      message['Source ID']
    ) {
      const DBInstanceIdentifier = message['Source ID'];
      const { DBInstances } = await this.rds.send(
        new DescribeDBInstancesCommand({
          DBInstanceIdentifier,
        })
      );
      if (DBInstances) {
        const description =
          DBInstances.find(
            (i) => i.DBInstanceIdentifier === DBInstanceIdentifier
          ) ?? {};
        const { PerformanceInsightsEnabled } = description;
        if (description && !PerformanceInsightsEnabled) {
          let complete = false;
          let attempts = 0;
          while (!complete && attempts < retryAttempts) {
            try {
              await this.rds.send(
                new ModifyDBInstanceCommand({
                  DBInstanceIdentifier,
                  EnablePerformanceInsights: true,
                })
              );
              complete = true;
            } catch (error) {
              console.error(error);
              if (
                error instanceof RDSServiceException &&
                (error as any).Code === 'InvalidDBInstanceState'
              ) {
                attempts += 1;
                await delay(retryDelay);
              } else {
                throw error;
              }
            }
          }
        }
      }
    }
  }

  protected async getTagsForCluster(
    dBInstanceIdentifier: string
  ): Promise<Array<Required<Tag>>> {
    const { DBInstances = [] } = await this.rds.send(
      new DescribeDBInstancesCommand({
        DBInstanceIdentifier: dBInstanceIdentifier,
      })
    );
    const [instance] = DBInstances;
    if (!instance) return [];
    const { DBClusterIdentifier } = instance;
    if (!DBClusterIdentifier) return [];
    const { DBClusters = [] } = await this.rds.send(
      new DescribeDBClustersCommand({
        DBClusterIdentifier,
      })
    );
    const [cluster] = DBClusters;
    return cluster?.TagList?.filter(this.isCompleteTag) ?? [];
  }

  protected async scale(
    reference: StackReference,
    minCapacity: number
  ): Promise<RegisterScalableTargetResponse> {
    const db = this.getClusterName(reference);
    const aasResponse = await this.aas.send(
      new DescribeScalableTargetsCommand({
        ServiceNamespace: 'rds',
        ResourceIds: [`cluster:${db}`],
      })
    );
    if (aasResponse?.ScalableTargets?.[0]) {
      const scalingTarget = aasResponse.ScalableTargets[0];
      const { ServiceNamespace, ResourceId, ScalableDimension } = scalingTarget;
      const response = await this.aas.send(
        new RegisterScalableTargetCommand({
          ResourceId,
          ScalableDimension,
          ServiceNamespace,
          MinCapacity: minCapacity,
        })
      );
      return response;
    }

    throw new Error(`unable to scale out db cluster ${db}`);
  }

  protected getClusterName(ref: StackReference): string {
    return ref === StackReference.a
      ? this.config.clusterNameA
      : this.config.clusterNameB;
  }

  protected getClusterPartnerName(ref: StackReference): string {
    return ref === StackReference.b
      ? this.config.clusterNameA
      : this.config.clusterNameB;
  }

  protected getClusterPartnerRef(ref: StackReference): StackReference {
    return ref === StackReference.a ? StackReference.b : StackReference.a;
  }

  private isCompleteTag(
    tag: Tag,
    _ix: number,
    _array: Tag[]
  ): tag is Required<Tag> {
    return Boolean(tag.Key && tag.Value);
  }
}
