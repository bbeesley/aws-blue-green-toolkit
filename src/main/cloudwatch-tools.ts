import {
  CloudWatchClient,
  DescribeAlarmsCommand,
  DisableAlarmActionsCommand,
  EnableAlarmActionsCommand,
} from '@aws-sdk/client-cloudwatch';
import type { AlarmIdentifiers, CloudWatchConfig } from './@types/index.js';
import { StackReference } from './constants.js';

/**
 * Toolkit for CloudWatch operations
 *
 * @export
 * @class CloudWatchTools
 */
export class CloudWatchTools {
  public cloudWatch: CloudWatchClient;

  /**
   * Creates an instance of CloudWatchTools.
   *
   * @param {CloudWatchConfig} config - Config describing the alarm pair
   * @memberof CloudWatchConfig
   */
  constructor(public config: CloudWatchConfig) {
    this.config = config;
    this.cloudWatch = new CloudWatchClient({ region: this.config.awsRegion });
  }

  /**
   * Disable all alarm actions
   *
   * @param {StackReference} reference - Reference to a subscription queue stack
   * @returns {Promise<void>}
   */
  public async disableAlarmsActions(reference: StackReference): Promise<void> {
    const alarms = await this.getAlarms(this.getAlarmIdentifiers(reference));
    await this.cloudWatch.send(
      new DisableAlarmActionsCommand({ AlarmNames: alarms })
    );
  }

  /**
   * Enable all alarm actions
   *
   * @param {StackReference} reference - Reference to a subscription queue stack
   * @returns {Promise<void>}
   */
  public async enableAlarmsActions(reference: StackReference): Promise<void> {
    const alarms = await this.getAlarms(this.getAlarmIdentifiers(reference));
    await this.cloudWatch.send(
      new EnableAlarmActionsCommand({ AlarmNames: alarms })
    );
  }

  private getAlarmIdentifiers(ref: StackReference): AlarmIdentifiers {
    return ref === StackReference.a
      ? this.config.alarmStackA
      : this.config.alarmStackB;
  }

  private async getAlarms(ref: AlarmIdentifiers): Promise<string[]> {
    const { MetricAlarms } = await this.cloudWatch.send(
      new DescribeAlarmsCommand({
        AlarmNamePrefix: ref.alarmPrefix,
      })
    );
    let alarms = MetricAlarms;
    if (!MetricAlarms)
      throw new Error(
        `unable to fetch cloudwatch metric alarms for ${JSON.stringify(ref)}`
      );
    const { alarmSuffix } = ref;
    if (alarmSuffix) {
      alarms = MetricAlarms.filter((r) => r.AlarmName?.endsWith(alarmSuffix));
    }

    return alarms?.map((r) => r.AlarmName).filter(Boolean) as string[];
  }
}
