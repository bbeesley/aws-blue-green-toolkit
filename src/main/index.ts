export * from './constants';
export { AuroraTools } from './AuroraTools';
