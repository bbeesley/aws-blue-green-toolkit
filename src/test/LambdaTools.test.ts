import * as AWS from 'aws-sdk';
import { LambdaTools, StackReference } from '../main';

const lambdaTools = new LambdaTools({
  awsRegion: 'eu-central-1',
  awsProfile: '555',
  environment: 'dev',
  namespace: 'fn',
  tags: {},
  lambdaNameA: 'fn-a-dev',
  lambdaNameB: 'fn-b-dev',
});

describe('LambdaTools', () => {
  describe('enableEventMapping', () => {
    it('calls updateEventSourceMapping with expected params', async () => {
      await lambdaTools.enableEventMapping(StackReference.b);
      expect(AWS.Lambda._listEventSourceMappings).toHaveBeenCalledWith({
        FunctionName: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev',
      });
      expect(AWS.Lambda._updateEventSourceMapping).toHaveBeenCalledWith({
        UUID: 'uuid',
        Enabled: true,
      });
    });
    it('calls updateEventSourceMapping with expected params and variant', async () => {
      const lambda = new LambdaTools({
        awsRegion: 'eu-central-1',
        awsProfile: '555',
        environment: 'dev',
        namespace: 'fn',
        tags: {},
        lambdaNameA: 'fn-a-dev',
        lambdaNameB: 'fn-b-dev',
        alias: 'live',
      });
      await lambda.enableEventMapping(StackReference.b);
      expect(AWS.Lambda._listEventSourceMappings).toHaveBeenCalledWith({
        FunctionName: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev:live',
      });
      expect(AWS.Lambda._updateEventSourceMapping).toHaveBeenCalledWith({
        UUID: 'uuid',
        Enabled: true,
      });
    });
  });
  describe('disableEventMapping', () => {
    it('calls updateEventSourceMapping with expected params', async () => {
      await lambdaTools.disableEventMapping(StackReference.b);
      expect(AWS.Lambda._listEventSourceMappings).toHaveBeenCalledWith({
        FunctionName: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev',
      });
      expect(AWS.Lambda._updateEventSourceMapping).toHaveBeenCalledWith({
        UUID: 'uuid',
        Enabled: false,
      });
    });
    it('calls updateEventSourceMapping with expected params and variant', async () => {
      const lambda = new LambdaTools({
        awsRegion: 'eu-central-1',
        awsProfile: '555',
        environment: 'dev',
        namespace: 'fn',
        tags: {},
        lambdaNameA: 'fn-a-dev',
        lambdaNameB: 'fn-b-dev',
        alias: 'live',
      });
      await lambda.disableEventMapping(StackReference.b);
      expect(AWS.Lambda._listEventSourceMappings).toHaveBeenCalledWith({
        FunctionName: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev:live',
      });
      expect(AWS.Lambda._updateEventSourceMapping).toHaveBeenCalledWith({
        UUID: 'uuid',
        Enabled: false,
      });
    });
  });
  describe('enableRule', () => {
    it('calls enableRule with expected params', async () => {
      await lambdaTools.enableRule(StackReference.b);
      expect(AWS.CloudWatchEvents._listRuleNamesByTarget).toHaveBeenCalledWith({
        TargetArn: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev',
      });
      expect(AWS.CloudWatchEvents._enableRule).toHaveBeenCalledWith({
        Name: AWS.CloudWatchEvents._eventsResponses.listRuleNamesByTarget
          .RuleNames[0],
      });
    });
    it('calls enableRule with expected params and variant', async () => {
      const lambda = new LambdaTools({
        awsRegion: 'eu-central-1',
        awsProfile: '555',
        environment: 'dev',
        namespace: 'fn',
        tags: {},
        lambdaNameA: 'fn-a-dev',
        lambdaNameB: 'fn-b-dev',
        alias: 'live',
      });
      await lambda.enableRule(StackReference.b);
      expect(AWS.CloudWatchEvents._listRuleNamesByTarget).toHaveBeenCalledWith({
        TargetArn: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev:live',
      });
      expect(AWS.CloudWatchEvents._enableRule).toHaveBeenCalledWith({
        Name: AWS.CloudWatchEvents._eventsResponses.listRuleNamesByTarget
          .RuleNames[0],
      });
    });
  });
  describe('disableRule', () => {
    it('calls disableRule with expected params', async () => {
      await lambdaTools.disableRule(StackReference.b);
      expect(AWS.CloudWatchEvents._listRuleNamesByTarget).toHaveBeenCalledWith({
        TargetArn: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev',
      });
      expect(AWS.CloudWatchEvents._disableRule).toHaveBeenCalledWith({
        Name: AWS.CloudWatchEvents._eventsResponses.listRuleNamesByTarget
          .RuleNames[0],
      });
    });
    it('calls disableRule with expected params and alias', async () => {
      const lambda = new LambdaTools({
        awsRegion: 'eu-central-1',
        awsProfile: '555',
        environment: 'dev',
        namespace: 'fn',
        tags: {},
        lambdaNameA: 'fn-a-dev',
        lambdaNameB: 'fn-b-dev',
        alias: 'live',
      });
      await lambda.disableRule(StackReference.b);
      expect(AWS.CloudWatchEvents._listRuleNamesByTarget).toHaveBeenCalledWith({
        TargetArn: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev:live',
      });
      expect(AWS.CloudWatchEvents._disableRule).toHaveBeenCalledWith({
        Name: AWS.CloudWatchEvents._eventsResponses.listRuleNamesByTarget
          .RuleNames[0],
      });
    });
  });
  describe('listEventSourceMappings', () => {
    it('calls listEventSourceMappings with expected params', async () => {
      const mappings = await lambdaTools.listEventSourceMappings(
        StackReference.b
      );
      expect(AWS.Lambda._listEventSourceMappings).toHaveBeenCalledWith({
        FunctionName: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev',
      });
      expect(mappings).not.toBeUndefined();
    });
    it('calls updateEventSourceMapping with expected params and variant', async () => {
      const lambda = new LambdaTools({
        awsRegion: 'eu-central-1',
        awsProfile: '555',
        environment: 'dev',
        namespace: 'fn',
        tags: {},
        lambdaNameA: 'fn-a-dev',
        lambdaNameB: 'fn-b-dev',
        alias: 'live',
      });
      const mappings = await lambda.listEventSourceMappings(StackReference.b);
      expect(AWS.Lambda._listEventSourceMappings).toHaveBeenCalledWith({
        FunctionName: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev:live',
      });
      expect(mappings).not.toBeUndefined();
    });
  });
  describe('createEventSourceMapping', () => {
    it('calls createEventSourceMapping with expected params', async () => {
      const mapping = await lambdaTools.createEventSourceMapping(
        StackReference.b,
        'some-arn',
        {
          StartingPosition: 'TRIM_HORIZON',
        }
      );
      expect(AWS.Lambda._createEventSourceMapping).toHaveBeenCalledWith({
        FunctionName: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev',
        EventSourceArn: 'some-arn',
        StartingPosition: 'TRIM_HORIZON',
      });
      expect(mapping).not.toBeUndefined();
    });
    it('calls updateEventSourceMapping with expected params and variant', async () => {
      const lambda = new LambdaTools({
        awsRegion: 'eu-central-1',
        awsProfile: '555',
        environment: 'dev',
        namespace: 'fn',
        tags: {},
        lambdaNameA: 'fn-a-dev',
        lambdaNameB: 'fn-b-dev',
        alias: 'live',
      });
      const mapping = await lambda.createEventSourceMapping(
        StackReference.b,
        'some-arn'
      );
      expect(AWS.Lambda._createEventSourceMapping).toHaveBeenCalledWith({
        FunctionName: 'arn:aws:lambda:eu-central-1:555:function:fn-b-dev:live',
        EventSourceArn: 'some-arn',
      });
      expect(mapping).not.toBeUndefined();
    });
  });
  describe('deleteEventMapping', () => {
    it('calls deleteEventMapping with expected params', async () => {
      const UUID = '091a66ea-1d4c-411c-97f0-039905401602';

      await lambdaTools.deleteEventMapping(UUID);
      expect(AWS.Lambda._deleteEventSourceMapping).toHaveBeenCalledWith({
        UUID,
      });
    });
  });
  describe('getLatestMetrics', () => {
    it('returns the latest metrics for the given lambda fn', async () => {
      const latestMetrics = await lambdaTools.getLatestMetrics(
        StackReference.b
      );
      expect(AWS.CloudWatch._getMetricData).toHaveBeenCalledTimes(1);

      expect(latestMetrics).toMatchSnapshot();
    });
  });
});
