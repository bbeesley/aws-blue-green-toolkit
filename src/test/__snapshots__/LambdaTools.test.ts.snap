// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LambdaTools deleteEventMapping calls deleteEventMapping with expected params 1`] = `
Object {
  "ConcurrentExecutions": Object {
    "latestMetricTimestamp": 2021-06-02T15:23:00.000Z,
    "latestMetricValue": 1,
    "stat": "Maximum",
  },
  "Duration": Object {
    "latestMetricTimestamp": 2021-06-02T15:23:00.000Z,
    "latestMetricValue": 10.4375,
    "stat": "Average",
  },
  "Errors": Object {
    "latestMetricTimestamp": 2021-06-02T15:23:00.000Z,
    "latestMetricValue": 4,
    "stat": "Sum",
  },
  "Invocations": Object {
    "latestMetricTimestamp": 2021-06-02T15:23:00.000Z,
    "latestMetricValue": 4,
    "stat": "Sum",
  },
  "IteratorAge": Object {
    "latestMetricTimestamp": 2021-06-02T15:23:00.000Z,
    "latestMetricValue": 5193039,
    "stat": "Maximum",
  },
  "Throttles": Object {
    "latestMetricTimestamp": 2021-06-02T15:23:00.000Z,
    "latestMetricValue": 0,
    "stat": "Sum",
  },
}
`;
